blackmarket/Managers/SessionManager.swift

```diff
@@ -0,0 +1,71 @@
+//
+//  SessionDataManager.swift
+//  ios-base
+//
+//  Created by Juan Pablo Mazza on 11/8/16.
+//  Copyright © 2016 Rootstrap Inc. All rights reserved.
+//
+
+import UIKit
+import Combine
+import SwiftUI
+
+internal class SessionManager: CurrentUserSessionProvider {
+    
+    var isSessionValidPublisher: AnyPublisher<Bool, Never> {
+        currentSessionPublisher.map { $0?.isValid ?? false }.eraseToAnyPublisher()
+    }
+
+    private var currentSessionPublisher: AnyPublisher<Session?, Never> {
+        userDefaults.publisher(for: \.currentSession).eraseToAnyPublisher()
+    }
+    
+    private var subscriptions = Set<AnyCancellable>()
+    private let userDefaults: UserDefaults
+    
+    static let SESSIONKEY = "ios-base-session"
+    
+    static let shared = SessionManager()
+        
+    init(userDefaults: UserDefaults = .standard) {
+        self.userDefaults = userDefaults
+    }
+    
+    private(set) var currentSession: Session? {
+        get {
+            userDefaults.currentSession
+        }
+        
+        set {
+            userDefaults.currentSession = newValue
+        }
+    }
+    
+    func deleteSession() {
+        userDefaults.removeObject(forKey: SessionManager.SESSIONKEY)
+    }
+    
+    func saveUser(session: Session) {
+        DispatchQueue.main.async {
+            self.userDefaults.currentSession = session
```

Comment:
we don't need to do this in the main thread

Reply:
This is the error I got

Reply:
Publishing changes from background threads is not allowed; make sure to publish values from the main thread (via operators like receive(on:)) on model updates.